[APM] Anomaly detection setup link with alert if job doesn't exist

ogupte opened the pull request:
Closes https://github.com/elastic/kibana/issues/70440 by adding a setup link to anomaly detection setting in the home header.

![anomaly-detection-setup-link-0](https://user-images.githubusercontent.com/1967266/87006002-cabd6b80-c174-11ea-9483-179d08ff22f0.gif)

![Screen Shot 2020-07-08 at 11 08 22 PM](https://user-images.githubusercontent.com/1967266/87006013-d27d1000-c174-11ea-9c83-c02714fbcc33.png)

mistic committed:
Revert "[i18n] integrate new translations + new i18n check #70193 (#70423)"

This reverts commit 2212beba6847051bd42f759ec2530d7c9521e64d.

ogupte committed:
Closes #70440 by adding a setup link to anomaly detection setting in the home header

lizozom committed:
Allow creating filters from fields with null values in discover (#70936)

* Fix bug #7189

* typo

* Test adjustments

* wait for load complete

* Fine tune test

* Update src/plugins/data/public/query/filter_manager/lib/generate_filters.ts

Co-authored-by: Lukas Olson <olson.lukas@gmail.com>

* Fix filtering by an array of nulls
Allow filtering by a non existing field in the doc
simplify flatten hit logic

Co-authored-by: Lukas Olson <olson.lukas@gmail.com>
Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

dej611 committed:
Improve failure reporting output for functional tests (#70954)

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

thomheymann committed:
Clone role button going to edit page (#70768)

* Clone role button going to edit page

* Added unit tests

* Fixed types

sorenlouv commented on x-pack/plugins/apm/public/components/shared/Links/apm/AnomalyDetectionSetupLink.tsx:
```diff
@@ -0,0 +1,79 @@
+export function AnomalyDetectionSetupLink() {
+  const { uiFilters, urlParams } = useUrlParams();
+  // check both uiFilters and urlParams for selected environment
+  const environment = uiFilters.environment || urlParams.environment;
```
This shouldn't be necessary. If it is we have a bug somewhere. `uiFilters` should read from `urlParams`

mistic committed:
Merge remote-tracking branch 'upstream/master'

mistic committed:
skip flaky suite (#71220)

flash1293 committed:
mitigate kql bug (#70712)

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

sorenlouv commented on x-pack/plugins/apm/public/components/shared/Links/apm/AnomalyDetectionSetupLink.tsx:
```diff
@@ -0,0 +1,79 @@
+    [],
+    { preservePreviousData: false }
+  );
+  const isFetchSuccess = status === FETCH_STATUS.SUCCESS;
+
+  const hasJobs = data.length > 0;
+  const hasJobForEnv = environment
+    ? data.some(({ environment: env }) => environment === env)
+    : true;
+
+  const showAlert = isFetchSuccess && (!hasJobs || !hasJobForEnv);
```
Maybe I'm missing something but afaict this can be simplified a lot:

```suggestion
const showAlert = data.some(job => environment === job.environment)
```

ogupte replied:
We would still need to check `isFetchSuccess` so it doesn't flash the alert prematurely:
```typescript
const showAlert = isFetchSuccess && !data.jobs.some((job) => environment === job.environment);
```

I do like how concise it is. I intentionally wrote it this way because I wanted the 2 conditions for showing an alert to be very clear in the code (1. there are no jobs OR 2. there are no jobs matching the current environment). It's not obvious with only the `some` call that it handles the case where no jobs exist. MDN docs even warns about this case specifically: "Caution: Calling this method on an empty array returns false for any condition!"

But I agree that less code written means less code to maintain. And we can always capture the various conditions in unit tests and code comments. Thanks for the suggestion.

sorenlouv replied:
I agree. We shouldn't aim to make things one liners if it hurts readability. In this case I think the shorter version _is_ the more readable though. 

And yes, I got it a little wrong. You adaptation is better:
```
const showAlert = isFetchSuccess && !data.jobs.some((job) => environment === job.environment);
```

rudolf committed:
Disable fields added from unknown saved object types (#70951)

* Allow disabled: false SO field mappings

* Disable fields for unknown SO types

* Update everyone else's docs ;)

* Address review comments

* Add unit tests for disableUnknownTypeMappingFields()

ogupte committed:
[APM] Service maps anomaly detection integration by environment (#70932)

* Closes #69480 & #70419.
- Adds anomaly detection integration to service maps backed by apm ML jobs per environment
- Loads transaction stats and anomalies for each transaction types
- Renders a selector in the popop to choose a transaction type to view stats

* - implements original anomaly detection integration design for service maps popover
- only aggregates transaction KPIs and anomaly scores for transaction.type = "request" or "page-load"
- supports environment filter 'All' option to display data from all APM anomaly detection jobs
- handle case where popover metrics don't exist for services outside the current environment filter

* fixes some CI errors

* Simplified messaging for service popop with not data in the current environment

* PR feedback, renamed max anomalies -> service anomalies including the file name

* - defines custom_settings.job_tags.apm_ml_version in ML job creation,
  then filters for it when returing valid APM ML jobs

* changes shape of of service anomalies from an array to a object keyed by serviceName

* removes the url encoding from ML job link href to how it was previously.

* PR feedback

* Popover no data state simplified:
- renders the "no data" message as plain text instead of in a callout
- hides the 'Anomaly detection' section if there is not anomaly data.

* Fixes filtering bug when user selects 'Environment: Not defined'. Now
filters properly by filtering for docs where service.environment does
not exist

* filters jobs fetched in the settings page by `job.custom_settings.job_tags.apm_ml_version`

* Fixed bad import from last upstream merge

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>
